net: lwm2m: Update security object to 1.1

jarlamsa opened the pull request:
Update LwM2M security object to version 1.1
Added Kconfig option to switch between object versions 1.0 and 1.1

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

Commit:
net: lwm2m: Update security object to 1.1

Add 1.1 version of the LWM2M security object.

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

jarlamsa commented:
@VeijoPesonen @jheiskan81 @SeppoTakalo  please review

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_obj_security.c:
```diff
@@ -34,8 +45,13 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#if defined(CONFIG_LWM2M_SECURITY_OBJECT_VERSION_1_1)
+#define SECURITY_MATCHING_TYPE_ID		13
+#define SECURITY_SNI_ID				14
+#define SECURITY_CERTIFICATE_USAGE_ID		15
```
Extra whitespace.

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_obj_security.c:
```diff
@@ -72,7 +96,14 @@ static struct lwm2m_engine_obj_field fields[] = {
 	OBJ_FIELD_DATA(SECURITY_SHORT_SERVER_ID, W_OPT, U16),
 	OBJ_FIELD_DATA(SECURITY_CLIENT_HOLD_OFF_TIME_ID, W_OPT, U32),
-	OBJ_FIELD_DATA(SECURITY_BS_SERVER_ACCOUNT_TIMEOUT_ID, W_OPT, U32)
+	OBJ_FIELD_DATA(SECURITY_BS_SERVER_ACCOUNT_TIMEOUT_ID, W_OPT, U32),
```
I don't know does those really matter but the following fields are defined as signed integers - S32s - by the specification

![image](https://user-images.githubusercontent.com/33285290/147248187-ee86292d-9143-4f3d-9aed-02eb0e50a7a7.png)

jarlamsa replied:
Good catch. First of those 2 are already in the version 1.0, but have probably seen no use as they aren't initialized at all in the object instance creation.

VeijoPesonen reviewed (changes requested):
The changes look good but it seems that there are potential minor bugs adjacent to your modifications.

Commit:
net: lwm2m: Security object resource correct type

Changed resource IDs 11 and 12 of the security object to use
signed integer as they are defined in the OMA specification.

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

jarlamsa commented:
Thank you @VeijoPesonen for your review comments. Those have been now addressed, please re-review.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_security.c:
```diff
@@ -46,8 +62,16 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
  * start with SECURITY_MAX_ID
+ *
+ * If using object version 1.1
+ * subtract MULTI resources because their counts include 0 resource (1)
```
I don't see this being applied in the actual code (line 71).

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_security.c:
```diff
@@ -138,6 +169,13 @@ static struct lwm2m_engine_obj_inst *security_create(uint16_t obj_inst_id)
+	INIT_OBJ_RES_OPTDATA(SECURITY_MATCHING_TYPE_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SECURITY_SNI_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SECURITY_CERTIFICATE_USAGE_ID, res[index], i, res_inst[index], j);
+	INIT_OBJ_RES_OPTDATA(SECURITY_DTLS_TLS_CIPHERSUITE_ID, res[index], i, res_inst[index], j);
```
I think this should use `INIT_OBJ_RES_MULTI_OPTDATA` since it's a multi-instance resource.

jarlamsa commented:
Thank you for your review comments @rlubos I think I have addressed those, could you please re-review.

rlubos reviewed (approved):
LGTM
